pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -352,3 +377,71 @@ func isMaster(machine mapiv1.Machine, client client.Client) bool {
 	}
 	return false
 }
+
+// isMachineDisruptionAllowed returns true if the provided MachineDisruptionBudget allows any disruption
+func isMachineDisruptionAllowed(mdb *healthcheckingv1alpha1.MachineDisruptionBudget, maxDisruptedMachinSize int) bool {
+	if mdb.Status.ObservedGeneration < mdb.Generation {
+		glog.Warningf("the machine disruption budget %s is still being processed by the server", mdb.Name)
+		return false
+	}
+	if mdb.Status.MachineDisruptionsAllowed < 0 {
+		glog.Warningf("the machine disruption budget %s MachineDisruptionsAllowed is negative", mdb.Name)
+		return false
+	}
+	if len(mdb.Status.DisruptedMachines) > maxDisruptedMachinSize {
+		glog.Warningf("the machine disruption budget %s DisruptedMachines map too big - too many deletions not confirmed by MDB controller", mdb.Name)
+		return false
+	}
+	if mdb.Status.MachineDisruptionsAllowed == 0 {
+		glog.Warningf("cannot delete machine as it would violate the machine's disruption budget %s", mdb.Name)
+		return false
+	}
+
+	return true
+}
+
+func (r *ReconcileMachineHealthCheck) decrementMachineDisruptionsAllowed(machineName string, mdb *healthcheckingv1alpha1.MachineDisruptionBudget) error {
+	mdb.Status.MachineDisruptionsAllowed--
+	if mdb.Status.DisruptedMachines == nil {
+		mdb.Status.DisruptedMachines = make(map[string]metav1.Time)
+	}
+	// MachineHealthCheck controller needs to inform the MDB controller that it is about to delete a machine
+	// so it should not consider it as available in calculations when updating MachineDisruptions allowed.
+	// If the machine is not deleted within a reasonable time limit MDB controller will assume that it won't
+	// be deleted at all and remove it from DisruptedMachines map.
```

Comment:
It's up to MHC to decide when a machine is healthy and when it's not. Thus, MDB controller can not make that decision and remove a machine from `mdb.Status.DisruptedMachines`. Also, object's status reflects object's recociler's opinion about object's current state. Thus, updating object status with two different controllers is not recommended. I would rather have MDB to be the only controller allowed to update MDB object's status. But again, MHC is responsible for deciding which machine is unhealthy and which machine needs to be re-mediated. So it's only up to MHC to modify `mdb.Status.DisruptedMachines` field.

Reply:
I can see that `DisruptedMachines` field raises a lot of questions, to be honest, for me, it also hard to imagine the situation when:
```
	// DisruptedMachines contains information about machines whose deletion was
	// processed by the API server but has not yet been observed by the MachineDisruptionBudget controller.
	// A machine will be in this map from the time when the API server processed the
	// deletion request to the time when the machine is seen by MDB controller
	// as having been marked for deletion (or after a timeout).
```
but I made this controller similar to pod distribution one, so I believe k8s guys had some issues with it. But we can consider removing `DisruptedMachines` in the next iteration, but for now, I prefer to leave it close as possible to the pod disruption controller to avoid some corner cases that we still did not encounter.